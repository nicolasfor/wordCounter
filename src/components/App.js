import React, { Component } from 'react';
import logo from '../assets/logo.svg';
import Dropzone from 'react-dropzone'
import TreeView from 'react-treeview';
import blackList from '../config/blackList'
import 'react-treeview/react-treeview.css'
class App extends Component {

  constructor(state)
  {
      super(state);
      this.state = {files:null,dictionary:null,plainText:null,toCollapse:false}
      this.handleClick = this.handleClick.bind(this)
      this.collapseAll = this.collapseAll.bind(this)
  }
  onDrop(files) {
    this.createDictionary(files)
  }
  async createDictionary(files)
  {
    //console.log("files",files);
    var dictionary = {};
    let promises=[];
    if(Array.isArray(files) && files.length>0)
    files.forEach((item,index)=>{
      //console.log("about to call");
        promises.push(this.readerPromise(item))
      })

    try{
      let results = await Promise.all(promises);
      if(results && results.length>0)
      {
        let plainText = results.join(' ');
        plainText.toLowerCase().split(' ')
        .forEach((word,index)=>{
          word= word.replace(/[^\w\s]/gi, '').replace(/\r?\n|\r/)
          if(word)
          {
            let validWord = blackList.find((item,index)=>item==word);
            if(!validWord )
            {
              let firstLetter = word[0];
              if(!dictionary[firstLetter])
                dictionary[firstLetter]={};
              if(!dictionary[firstLetter][word])
              dictionary[firstLetter][word]=1;
              else
              dictionary[firstLetter][word]+=1;
            }
          }
        })
       console.log("dictionary",dictionary);
        this.setState({dictionary,plainText,collapsedBookkeeping:Object.keys(dictionary).map((item,index)=>true)})
      }
      
    }

    catch(error){
      console.log("error",error);
      alert("Error: "+error)
    }
  }

  readerPromise(file)
  {
    return new Promise((resolve,reject)=>{
      var reader = new FileReader();
      reader.onload = ()=> {
        console.log("reader.result",reader.result);
        if(file.type.match('text/plain'))
          resolve(reader.result);
        else
          reject(file.name+" is not a valid format")
      }
      reader.onerror = ()=> {
        reject("Opps something happened!");
      }
      reader.readAsText(file)
    })
  }

  handleClick(i) {
    let [...collapsedBookkeeping] = this.state.collapsedBookkeeping;
    collapsedBookkeeping[i] = !collapsedBookkeeping[i];
    let index = collapsedBookkeeping.findIndex((item,index)=>item==false?true:false)
    this.setState({collapsedBookkeeping: collapsedBookkeeping,toCollapse:index>=0?true:false});
  }

  collapseAll() {
    let {toCollapse} = this.state
    this.setState({
      collapsedBookkeeping: this.state.collapsedBookkeeping.map(() => toCollapse),
      toCollapse:!toCollapse
    });
  }

  render() {
    let {dictionary,plainText,toCollapse,collapsedBookkeeping} = this.state
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <h1 className="App-title">Verification and Validation</h1>
        </header>
        <div className="App-intro">
          <div className="card">
            <div className="card-header">
              <b>Text Processing:</b>
            </div>
            {dictionary && plainText &&
            <div className="card-body">
              <div className="container">
                <div className="row">
                  <div className="col-sm">
                    {plainText}
                  </div>
                  <div className="col-sm tree-view-custom">
                  <button type="button" onClick={this.collapseAll} className="btn btn-primary">{toCollapse?"Collapse all":"Expand all"}</button>
                    {Object.keys(dictionary).sort().map((node, i) => {
                      // Let's make it so that the tree also toggles when we click the
                      // label. Controlled components make this effortless.
                      const label =
                        <span className="node" onClick={this.handleClick.bind(null, i)}>
                          Letter: {node}
                        </span>;
                      return (
                        <TreeView
                          key={i}
                          nodeLabel={label}
                          collapsed={collapsedBookkeeping[i]}
                          onClick={this.handleClick.bind(null, i)}>
                          {Object.keys(dictionary[node]).map(entry => <div className="info" key={entry}>{entry+": "+dictionary[node][entry]}</div>)}
                        </TreeView>
                      );
                    })}
                  </div>
                </div>
              </div>
            </div>
            }
            <div className="card-footer text-muted">
              <Dropzone className="dropfiles" onDrop={this.onDrop.bind(this)}>
              <p>Try dropping some files here, or click to select files to upload.</p>
            </Dropzone>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default App;
