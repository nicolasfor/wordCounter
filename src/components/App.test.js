import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Dropzone from 'react-dropzone'

describe(App, () => {
	const plainText='@you i&%&% an $%construct a f%ile...'
	const parts = [
	  new Blob([plainText], {type: 'text/plain'})
	];

	// Construct a file
	const file = new File(parts, 'sample.txt', {
	    lastModified: new Date(0),
	    type: "text/plain"
	});

	const files = [file]

	const parts2 = [
	  new Blob([plainText], {type: 'mov'})
	];
	const badFile = new File(parts2, 'sample.mov', {
	    lastModified: new Date(0),
	    type: "mov"
	});

	it('renders without crashing', () => {
	  const div = document.createElement('div');
	  ReactDOM.render(<App />, div);
	  ReactDOM.unmountComponentAtNode(div);
	});

	it("renders and matches our snapshot", () => {
	    const component = renderer.create(<App />);
	    const tree = component.toJSON();
	    expect(tree).toMatchSnapshot();
	  });

	it('should called createDictionary', () => {
	    const componentWrapper   = shallow(<App/>);
	    const component          = componentWrapper.instance();

	    const spy = jest.spyOn(component, "createDictionary");
	    componentWrapper.find(Dropzone).simulate('drop', files)
	    expect(spy).toHaveBeenCalledTimes(1)
	    expect(spy).toHaveBeenCalledWith(files)
	});

	it('should call readerPromise file size times', async () => {
	    const componentWrapper = shallow(<App />);
	    const component = componentWrapper.instance();
		const spy = jest.spyOn(component, "readerPromise");
		await component.createDictionary(files)
		expect(spy).toHaveBeenCalledTimes(files.length)
	});

	it('should detect file types not allowed', async () => {
	    const componentWrapper = shallow(<App />);
	    const component = componentWrapper.instance();

		expect.assertions(1);
		try {
		    await component.readerPromise(badFile)
		} catch (e) {
		    expect(e).toEqual(badFile.name+" is not a valid format");
		}
	});

	it('Plain text should keep as original', async () => {
	    const componentWrapper = shallow(<App />);
	    const component = componentWrapper.instance();
		await component.createDictionary(files)
		expect(componentWrapper.state().plainText).toEqual(plainText);
	});

	it('dictionary should populate correctly deleting black list words', async () => {
	    const componentWrapper = shallow(<App />);
	    const component = componentWrapper.instance();
		await component.createDictionary(files)
		let obj ={"c": {"construct": 1}, "f": {"file": 1}}
		expect(componentWrapper.state().dictionary).toEqual(obj);
	});

	it('Collapsable state should be false unless one item is clicked', async () => {
	    const componentWrapper = shallow(<App />);
	    const component = componentWrapper.instance();
		await component.createDictionary(files)
		expect(componentWrapper.state().toCollapse).toEqual(false);
		component.handleClick(0);
		expect(componentWrapper.state().toCollapse).toEqual(true);
	});
});