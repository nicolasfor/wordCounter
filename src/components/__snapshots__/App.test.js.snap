// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class App extends _react.Component {

  constructor(state)
  {
    super(state);
    this.state = { files: [], dictionaries: [] };
  }
  onDrop(files) {
    let newFiles = this.state.files.concat(files);
    this.setState({ files: newFiles });
    this.createDictionary(files);
  }
  async createDictionary(files)
  {var
    dictionaries = this.state.dictionaries;
    let promises = [];
    if (Array.isArray(files) && files.length > 0)
    files.forEach((item, index) => {
      console.log("item", item);
      promises.push(this.readerPromise(item));
    });

    try {
      let results = await Promise.all(promises);

      results.forEach((plainText, index) => {
        if (plainText)
        {
          let dictionary = {};
          plainText.toLowerCase().split(' ').
          forEach((word, index) => {
            word = word.replace(/[^\\w\\s]/gi, '').replace(/\\r?\\n|\\r/, '').replace(/\\s/g, '');
            if (word && word.length > 0)
            {
              let validWord = _blackList2.default.find((item, index) => item == word);
              if (!validWord)
              {
                let firstLetter = word[0];
                if (!dictionary[firstLetter])
                dictionary[firstLetter] = {};
                if (!dictionary[firstLetter][word])
                dictionary[firstLetter][word] = 1;else

                dictionary[firstLetter][word] += 1;
              }
            }
          });
          dictionaries.push({ dictionary, plainText, name: files[index].name });
        }
      });
      this.setState({ dictionaries });
    }
    catch (error) {
      console.log("error", error);
      alert("Error: " + error);
    }
  }

  readerPromise(file)
  {
    return new Promise((resolve, reject) => {
      var reader = new FileReader();
      reader.onload = () => {
        console.log("reader.result", reader.result);
        if (file.type.match('text/plain'))
        resolve(reader.result);else

        reject(file.name + " is not a valid format");
      };
      reader.onerror = () => {
        reject("Opps something happened!");
      };
      reader.readAsText(file);
    });
  }

  render() {let
    dictionaries = this.state.dictionaries;
    return (
      _react2.default.createElement('div', { className: 'App', __source: { fileName: _jsxFileName, lineNumber: 87 }, __self: this },
        _react2.default.createElement('header', { className: 'App-header', __source: { fileName: _jsxFileName, lineNumber: 88 }, __self: this },
          _react2.default.createElement('img', { src: _logo2.default, className: 'App-logo', alt: 'logo', __source: { fileName: _jsxFileName, lineNumber: 89 }, __self: this }),
          _react2.default.createElement('h1', { className: 'App-title', __source: { fileName: _jsxFileName, lineNumber: 90 }, __self: this }, 'Verification and Validation')),

        _react2.default.createElement('div', { className: 'App-intro', __source: { fileName: _jsxFileName, lineNumber: 92 }, __self: this },
          dictionaries.map((dictionary, index) => _react2.default.createElement(_Dictionary2.default, { key: index, dictionary: dictionary, __source: { fileName: _jsxFileName, lineNumber: 93 }, __self: this })),
          _react2.default.createElement('div', { className: 'card-footer text-muted', __source: { fileName: _jsxFileName, lineNumber: 94 }, __self: this },
            _react2.default.createElement(_reactDropzone2.default, { className: 'dropfiles', onDrop: this.onDrop.bind(this), __source: { fileName: _jsxFileName, lineNumber: 95 }, __self: this },
              _react2.default.createElement('p', { __source: { fileName: _jsxFileName, lineNumber: 96 }, __self: this }, 'Try dropping some files here, or click to select files to upload.'))))));





  }} renders and matches our snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Verification and Validation
    </h1>
  </header>
  <div
    className="App-intro"
  >
    <div
      className="card-footer text-muted"
    >
      <div
        aria-disabled={false}
        className="dropfiles"
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <p>
          Try dropping some files here, or click to select files to upload.
        </p>
        <input
          accept={undefined}
          autoComplete="off"
          disabled={false}
          multiple={true}
          onChange={[Function]}
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 0.00001,
              "pointerEvents": "none",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          type="file"
        />
      </div>
    </div>
  </div>
</div>
`;
